File: tools/runMopac.py
#!/usr/bin/env python
'''
#=============================================================================
#     FileName: runMopac.py
#         Desc: 
#       Author: jlpeng
#        Email: jlpeng1201@gmail.com
#     HomePage: 
#      Created: 2014-01-03 16:41:09
#   LastChange: 2014-02-18 14:33:18
#      History:
#=============================================================================
'''
import sys
import os
import glob
import pybel
import openbabel as ob

global mopac7
if sys.platform == "win32":
    mopac7 = "mopac7.exe"
else:
    mopac7 = "mopac7"

def main(argv=sys.argv):
    if len(argv) < 2:
        print "\n  Usage: runMopac7.py mols[...]\n"
        print "  Attention:"
        print "  1. molecule title is needed for each input molecule, and"
        print "     it will be used to name the output file!"
        print "  2. if the input molecule is charged or radical, you may need"
        print "     to add DOUBLET to the keyword list passed to Mopac"
        print "  3. .mopout file will be generated in the same dir as input!"
        print "  4. by default, the program 'mopac7' can be found in PATH. Otherwise,"
        print "     you need to specify the global variable 'mopac7' to tell the script"
        print "     where to find mopac7!"
        print "  5. the input file name can contain simple shell-style wildcards!"
        print ""
        sys.exit(1)

    for candidate in argv[1:]:
        names = glob.glob(candidate)
        for name in names:
            run(name)

def genMopFile(mol):
    conv = ob.OBConversion()
    conv.SetOutFormat("mop")
    if os.sep in mol.title:
        mop_name = mol.title[mol.title.rfind(os.sep)+1:] + ".mop"
    else:
        mop_name = mol.title + ".mop"
    if not mol.OBMol.HasHydrogensAdded():
        mol.addh()
    if not mol.OBMol.Has3D():
        mol.make3D()
    key = "ESP MMOK VECTORS BONDS PI PRECISE ENPART AM1 GNORM=0.01 EF XYZ MULLIK +\nCharge=%d"%mol.charge
    conv.AddOption("k",conv.OUTOPTIONS,key)
    status = conv.WriteFile(mol.OBMol, mop_name)
    if not status:
        print "Error: failed to write mop file for "%mol.title
        return ""
    else:
        return mop_name

def runMopac(mop_name):
    lines = open(mop_name,"r").readlines()
    fi_tmp = open("FOR005","w")
    fi_tmp.writelines(lines)
    fi_tmp.close()
    basename = mop_name[:mop_name.rfind(".")]
    mopout_name = basename+".mopout"
    cmd = "%s > %s"%(mopac7,mopout_name)
    status = os.system(cmd)
    elim_list=["FOR005","FOR009","FOR010","FOR011","FOR012","SHUTDOWN"]
    for i in elim_list:
        if os.path.exists(i):
            os.unlink(i)
    if status:
        print "Error: failed to run Mopac7 for molecule \"%s\""%basename
        return False
    lines = open(mopout_name,"r").readlines()
    if "DONE" not in lines[-1]:
        print "Error: failed to run Mopac7 for molecule \"%s\""%basename
        return False
    else:
        os.unlink(mop_name)
        print "\"%s\" has been generated!"%mopout_name
        return True


def run(name):
    dirname = os.path.dirname(name)
    basename = os.path.basename(name)
    _format = basename[basename.rfind(".")+1:]
    if _format not in ("mol","mol2","sdf"):
        print "Error: invalid format %s, only \"mol, mol2, and sdf\" supported"%_format
        return
    prev_path = os.getcwd()
    if dirname != "":
        os.chdir(dirname)
    error_list = []
    print "to optimize molecule(s) from",name
    for mol in pybel.readfile(_format,basename):
        print "> molecule titled \"%s\""%mol.title
        mop_name = genMopFile(mol)
        if mop_name == "":
            error_list.append(mol.title)
            continue
        if not runMopac(mop_name):
            error_list.append(mol.title)
    if dirname != "":
        os.chdir(prev_path)
    
    if len(error_list) != 0:
        print "\nAttention: failed to run Mopac for the following molecules:"
        for item in error_list:
            print item
    else:
        print "\nAll finished successfully"

main()

File: tools/runMopac.cpp
/*=============================================================================
#     FileName: runMopac.cpp
#         Desc: 
#       Author: jlpeng
#        Email: jlpeng1201@gmail.com
#     HomePage: 
#      Created: 2014-02-19 18:50:29
#   LastChange: 2014-03-20 17:10:19
#      History:
=============================================================================*/
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <vector>
#include <cstdlib>
#include <cstring>
#include <openbabel/mol.h>
#include <openbabel/obconversion.h>
#include <openbabel/builder.h>
#include <openbabel/forcefield.h>


using namespace std;
using namespace OpenBabel;

vector<string> runMopac(string &infile, string &mopac7exe, string &keywords);

void exit_with_help(char *name)
{
    cerr << endl << "Usage"<< endl
        << "  " << name << " [options]" << endl
        << endl << "[options]" << endl
        << "  -m inmol" << endl
        << "     all formats OpenBabel supported" << endl
        << "  -l inlist" << endl
        << "     file, each line should be a molecule, e.g. F:\\mols\\1.mol" << endl
        << "  --mopac dir" << endl
        << "     directory where mopac7.exe could be found" << endl
        << "     e.g. F:\\software" << endl
        << "     default: mopac7.exe could be found in the current dir or in PATH" << endl
        << "  --key words" << endl
        << "     additional keywords passed to mopac7. separated by ';'" << endl
        << "     e.g. DOUBLET;BOND" << endl
        << endl << "Attention:" << endl
        << "  1. output files will be named as 'basename_i.mopout'" << endl
        << "  2. output files will be generated in the same directory as its input" << endl
        << "  3. keywords being used by default:" << endl
        << "     ESP MMOK VECTORS BONDS PI PRECISE ENPART AM1 GNORM=0.01" << endl
        << "     EF XYZ MULLIK Charge" << endl
        << endl;
    exit(EXIT_FAILURE);
}

int main(int argc, char *argv[])
{
    if(argc<3 || argc>9)
        exit_with_help(argv[0]);

    string molfile("");
    string mollist("");
    string mopacdir("");
    string keywords("");
    string tmp_keywords("");
    for(int i=1; i<argc; ++i) {
        if(strcmp(argv[i], "-m") == 0)
            molfile = argv[++i];
        else if(strcmp(argv[i],"-l") == 0)
            mollist = argv[++i];
        else if(strcmp(argv[i],"--mopac") == 0)
            mopacdir = argv[++i];
        else if(strcmp(argv[i],"--key") == 0)
            tmp_keywords = argv[++i];
        else {
            cerr << "Error: invalid option " << argv[i] << endl;
            exit(EXIT_FAILURE);
        }
    }

    if(molfile=="" && mollist=="") {
        cerr << "Error: at least one of '-m -l' is needed!" << endl;
        exit(EXIT_FAILURE);
    }
    
    string mopac7exe;
    if(mopacdir == "")
        mopac7exe = "mopac7.exe";
    else {
        if(mopacdir[mopacdir.size()-1] == '\\')
            mopac7exe = mopacdir + "mopac7.exe";
        else
            mopac7exe = mopacdir + "\\mopac7.exe";
    }

    string::size_type i=0;
    while(i < tmp_keywords.size()) {
        int count = 0;
        int start_pos = i;
        while(i<tmp_keywords.size() && tmp_keywords[i]!=';') {
            ++count;
            ++i;
        }
        keywords += (" " + tmp_keywords.substr(start_pos, count));
    }

    vector<string> fail_list;

    if(molfile != "") {
        vector<string> tmp = runMopac(molfile, mopac7exe, keywords);
        for(vector<string>::iterator i=tmp.begin(); i!=tmp.end(); ++i)
            fail_list.push_back(*i);
    }
    if(mollist != "") {
        ifstream inf(mollist.c_str());
        if(!inf) {
            cerr << "Error: failed to open file " << mollist << endl;
            exit(EXIT_FAILURE);
        }
        string line;
        while(getline(inf,line)) {
            if(line.size() == 0)
                continue;
            if(line[0] == '#')
                continue;
            vector<string> tmp = runMopac(line, mopac7exe, keywords);
            for(vector<string>::iterator i=tmp.begin(); i!=tmp.end(); ++i)
                fail_list.push_back(*i);
        }
        inf.close();
    }

    if(!fail_list.empty()) {
        cout << endl << "failed to run mopac7 for the following files:" << endl;
        for(vector<string>::iterator i=fail_list.begin(); i!=fail_list.end(); ++i)
            cerr << "  " << *i << endl;
    } else
        cout << endl << "All finished!" << endl;

    return 0;
}

static void make3D(OBMol &mol)
{
    // reference: pybel.py::Molecule::make3D
    OBBuilder builder;
    builder.Build(mol);
#if defined(_MSC_VER)
    OBForceField* ff = (OBForceField *)OBPlugin::GetPlugin("forcefields", "MMFF94");
#else
    OBForceField *ff = OBForceField::FindType("mmff94");
#endif
    if(ff) {
        ff->SetLogLevel(OBFF_LOGLVL_LOW);
        bool success = ff->Setup(mol);
        if(!success) {
            cout << "   can't set the forcefield for molecule " << mol.GetTitle() << endl;
            exit(EXIT_FAILURE);
        }
        ff->SteepestDescent(50);
        ff->GetCoordinates(mol);
    } else {
        cout << "   Can't load force field: MMFF94!!!" << endl;
        exit(EXIT_FAILURE);
    }
}


vector<string> runMopac(string &infile, string &mopac7exe, string &keywords)
{
    cout << ">>> to process file " << infile;
    vector<string> fail_list;
    // 1. get format and base name
    string::size_type i = infile.rfind(".");
    if(i == string::npos) {
        cout << endl << "Error: can't recognize the format of " << infile << endl;
        fail_list.push_back(infile);
        return fail_list;
    }
    string format = infile.substr(i+1);
    string basic_name = infile.substr(0,i);
    // 2. read and process molecles
    OBMol mol;
    OBConversion conv;
    conv.SetInAndOutFormats(format.c_str(), "mop");
    bool next = conv.ReadFile(&mol, infile.c_str());
    int count = 0;
    while(next) {
        ++count;
        cout << endl << "    " << count << "th molecule... ";
        ostringstream os;
        os << basic_name << "_" << count << ".mopout";
        string mopout_name = os.str();
        // 2.1 add hydrogens or generate 3D coordinates
        if(!mol.HasHydrogensAdded())
            mol.AddHydrogens();
        if(!mol.Has3D())
            make3D(mol);
        // 2.2 generate FOR005
        char keys[256];
        sprintf(keys,
                "ESP MMOK VECTORS BONDS PI PRECISE ENPART AM1 GNORM=0.01 EF XYZ MULLIK +\nCharge=%d%s",
                mol.GetTotalCharge(), keywords.c_str());
        conv.AddOption("k",OBConversion::OUTOPTIONS, keys);
        ofstream outf("FOR005");
        outf << conv.WriteString(&mol);
        outf.close();
        // 2.3 run Mopac
        string cmd;
        cmd = mopac7exe + "> " + mopout_name;
        int status = system(cmd.c_str());
        if(status) {
            fail_list.push_back(mopout_name);
            cout << "failure";
        }
        else {
            ifstream inf(mopout_name.c_str());
            string line, lastline;
            while(getline(inf,line)) lastline=line;
            if(lastline.find("DONE") == string::npos) {
                fail_list.push_back(mopout_name);
                cout << "failure";
            } else
                cout << "success";
            inf.close();
        }
        system("del FOR005 FOR009 FOR010 FOR011 FOR012 SHUTDOWN");
        // next molecule
        next = conv.Read(&mol);
    }
    if(count == 0) {
        cout << "NO molecule being found!";
        fail_list.push_back(infile);
    }
    cout << endl;

    return fail_list;
}

